torch/onnx/_internal/fx/onnxfunction_dispatcher.py

```diff
@@ -494,11 +494,20 @@ def perfect_match_inputs(
                 # of this input defined in the OpSchema, we know the function
                 # and the input are not compatible
                 return False
-        # TODO: attributes dtype is not supported right now, so we just check their existence.
-        if set(function_attributes) != self.attributes:
+        # Check attributes numbers are the same
+        if set(function_attributes) != set(self.attributes):
             # If the attributes of the OpSchema and the attributes don't match,
             # we know the function and the input are not compatible
             return False
+        # Check attribute dtypes
+        for attribute_name, attribute in function_attributes.items():
+            attribute_onnx_type = fx_type_utils.from_python_type_to_onnx_attribute_type(
+                type(attribute)
+            )
+            if attribute_onnx_type != self.attributes[attribute_name].type:
+                # If the attribute type of the OpSchema and the attribute type don't match,
```

Comment:
Will this reasoning be visible in diagnostic?

Reply:
I updated the rule description to include attribute dtype

Reply:
When the input is int and the signature is float. Do we consider them compatible? I tend to think we should

Reply:
> I updated the rule description to include attribute dtype

I was thinking about accumulating very fine grained diagnostic message. Like adding `this is not perfect match because attribute {name} has {actual_type} vs {expected_type}.` into additional messages. Similarly for argument checks.

Doesn't need to be in this PR if these are not done before.

Reply:
> When the input is int and the signature is float. Do we consider them compatible? I tend to think we should

Should we make this exception?

Reply:
> > I updated the rule description to include attribute dtype
> 
> I was thinking about accumulating very fine grained diagnostic message. Like adding `this is not perfect match because attribute {name} has {actual_type} vs {expected_type}.` into additional messages. Similarly for argument checks.
> 
> Doesn't need to be in this PR if these are not done before.

https://github.com/pytorch/pytorch/issues/105174

Reply:
> > When the input is int and the signature is float. Do we consider them compatible? I tend to think we should
> 
> Should we make this exception?

Python / PyTorch treats int and float pretty loosely so I think it’s reasonable.  PyTorch’s Scalar type can be an int or float. We in the graph building process will convert int to float if the attribute type is float

Reply:
So I think it’s reasonable to be able to match. Otherwise we can also fix that in a pass?

Reply:
If there is an overload relying on this attribute dtype (int or float), this would be an issue.

Reply:
That's true. I guess the fuzzy match logic can handle it for now then?